### ISSUE TITLE
GitLab data collection 403 forbidden error on v0.10.1

### ISSUE THREAD

**Issue by hezyin** [apache/incubator-devlake#1937]
## Describe the bug

A seed user (Milan from CROZ) reported a GitLab data collection issue. He was able to successfully collect for one GitLab project and failed for the second project he tried.

The pipeline failed at progress 52.17% (took 1 min) with a 403 forbidden error message from GitLab. Since he did successfully collect one project and made sure his token has sufficient permission for all projects, Milan suspected the error is due to the absence of certain data for the failed project. This hypothesis is based on that GitLab sometimes gives a 403 forbidden error message when there's no data, see below for an example:

Visit
```
https://gitlab.croz.net/api/v4/projects/3060/jobs 
```

and get the following response:
```
{
    message: "403 Forbidden"
}
```
Project 3060 doesn't have any jobs.

## How to reproduce

* DevLake version: v0.10.1
* GitLab version: On-prem v14.10.2

**Comment by mappjzc** [comment]
We have try to run the api  `/api/v4/projects/3060/jobs` 
With the no project, it got 404 ERROR but not 403.
![image](https://user-images.githubusercontent.com/2921251/169303607-d4131aaa-c41d-4401-a056-731cb2c07404.png)
With the project but with out any jobs data, it got an empty array ,but not 403.
![image](https://user-images.githubusercontent.com/2921251/169303783-55162193-629e-4db5-8c85-69f5555ea8a8.png)

We are testing on the gitlab verson what your said.
![image](https://user-images.githubusercontent.com/2921251/169304647-a00cf88d-1043-4cec-9205-bf21edfe9795.png)


We  had also do the same testing at the version `GitLab Community Edition [13.5.3]` and gitlab.com ( version : `GitLab Enterprise Edition 15.0.0-pre [a81a1e8bf2e]` )

We havn't see the 403 Error at not only empty data but also no project.

Perhaps Milan had run the two project gitlab collect in the same time and it out of ratelimit so be banned by his gitlab?
Please tell us more detail about this issue.

**Comment by mjovanovic0** [comment]
Hi,

Now I receive 404 instead of 403. Access token is valid, connection i Online but when running the pipeline it receive 404.

<img width="1702" alt="Screenshot 2022-06-06 at 11 32 30" src="https://user-images.githubusercontent.com/6703229/172135517-27fdfc67-b73b-4e9b-b3e3-3cb40abfc9b5.png">
<img width="403" alt="Screenshot 2022-06-06 at 11 32 43" src="https://user-images.githubusercontent.com/6703229/172135534-5f1e6412-4793-4291-9545-d3ac99202421.png">

[devlake.log](https://github.com/apache/incubator-devlake/files/8843376/devlake.log)

**Comment by klesh** [comment]
Judge by the name, did this project contains any code? commits?

**Comment by mjovanovic0** [comment]
No, this project have no code. Even in config of the GitLab project, Repository feature is disabled.

![6F84AE18-0F43-443A-90D1-4075C518BBF0](https://user-images.githubusercontent.com/6703229/172460903-4331506f-4cbf-4402-a866-32bcf8e616c3.png)
![AEE471D9-796C-4CA2-84E2-095DF9018C84](https://user-images.githubusercontent.com/6703229/172460923-c911ea24-b0d3-408b-a184-765bfeb34629.png)

**Comment by hezyin** [comment]
@mjovanovic0 Hi Milan, another user reported the same 403 error message and we found out that it's because their project's CI/CD is disabled. This may explain why you got a 403 error message last time (and please do let us know if that wasn't the case, then we'll need to dig deeper into it).

I suspect the 404 error is due to the project's "Repository" functionality being turned off, but we'll try to reproduce it and get back to you. 

Thanks for following up on this issue.

**Comment by klesh** [comment]
@mappjzc would you take a look at this?

**Comment by mappjzc** [comment]
I have closed all Repository.
And the `api/v4/projects/37` can still work.
![image](https://user-images.githubusercontent.com/2921251/172656719-c421fbea-c116-4535-8099-ff143abe094c.png)
But `api/v4/projects/37/repository/commits` got the 404
![image](https://user-images.githubusercontent.com/2921251/172656474-37301e9d-a5d9-4df0-8b60-e1c60eabc6ae.png)

so it might be Repository setting not good.

**Comment by hezyin** [comment]
@mjovanovic0 Hi Milan, from @mappjzc's test, it seems it's indeed possible to get a 404 response from a GitLab project whose Repository functionality is turned off. 

This issue can be circumvented by enabling the Repository functionality in your repo as a temporary measure. In the long term, we'll need to figure out how to differentiate the real 404/403 error messages from false alarms when certain functionality is turned off.

Finally, if the GitLab project you tried to collect only has issues, then you may want to wait for our v0.11 release as v0.10.1 doesn't collect GitLab issues. v0.11 is in the release process and should be available next week. Thanks for your patience:)

**Comment by klesh** [comment]
Will be addressed by #2065
